Add `lu` decomposition for `Tensor`s

jofrevalles opened the pull request:
### Summary
This PR adds the new `lu` function for tensors, extending the `LinearAlgebra.lu` function (resolve #26). The new `lu` function returns the LU decomposition of a `Tensor`, where the tensor can be recovered by contracting the permutation tensor `P`, the tensor `L`, and tensor `U`. The tensors `L` and `U` are reshaped versions of the original lower and upper triangular matrices obtained during the decomposition process, respectively.

This implementation is inspired by the LU decomposition in the `scipy` library, as it returns the permutation tensor `P` allowing the original tensor `A` to be recovered with the contraction `A = P * L * U`. This contrasts with `LinearAlgebra`, where the permutation vector `p` is returned, and the original matrix can be recovered with `P' * A = L * U` (where `P'` is the permutation matrix built from `p`).

Please let me know if there are any concerns or issues with extending the `LinearAlgebra` library in this manner.

We have also added tests for this new function.

### Example
A usage example of the lu function:
```julia
julia> using Tenet; using LinearAlgebra; using Test

julia> tensor = Tensor(rand(4, 4, 4), (:i, :j, :k))
4×4×4 Tensor{Float64, 3, Array{Float64, 3}}:
...

julia> P, L, U = lu(tensor, left_inds = labels(tensor)[1:2])
...

julia> @test contract(contract(P, L), U) ≈ tensor
Test Passed
```

mofeing commented:
@jofrevalles I've updated the code to the recent changes. Also, I've refactored the SVD and QR factorizations. Would you mind giving a quick view?

jofrevalles commented:
> @jofrevalles I've updated the code to the recent changes. Also, I've refactored the SVD and QR factorizations. Would you mind giving a quick review?

I have checked it and everything looks good to me!
